src/main/kotlin/com/terraformation/backend/seedbank/db/AccessionStore.kt

```diff
@@ -858,4 +868,43 @@ class AccessionStore(
   fun getClock(accessionId: AccessionId): Clock {
     return clock.withZone(parentStore.getEffectiveTimeZone(accessionId))
   }
+
+  /**
+   * Runs a function on each accession in the database. Usually used for data migrations.
+   *
+   * Must be called as the system user.
+   *
+   * Calls [func] in a separate transaction for each accession, with a lock held on the accession's
+   * database row to prevent multiple server instances from processing the same accession at the
+   * same time.
+   *
+   * [func] should update the accession such that [condition] no longer matches it.
+   */
+  @Suppress("UNUSED")
+  fun forEachAccession(condition: Condition, func: (AccessionModel) -> Unit) {
+    var nextAccessionId: AccessionId? = null
+
+    if (currentUser().userType != UserType.System) {
+      throw AccessDeniedException("Migrations must run as system user")
+    }
+
+    do {
+      // Walk through accessions one at a time, locking each one since this migration could be
+      // running on multiple server instances at once.
+      dslContext.transaction { _ ->
+        nextAccessionId =
+            dslContext
+                .select(ACCESSIONS.ID)
+                .from(ACCESSIONS)
+                .where(condition)
+                .and(nextAccessionId?.let { ACCESSIONS.ID.gt(it) } ?: DSL.trueCondition())
+                .limit(1)
+                .forUpdate()
+                .skipLocked()
+                .fetchOne(ACCESSIONS.ID)
+
+        nextAccessionId?.let { accessionId -> func(fetchOneById(accessionId)) }
```

Comment:
should we be catching exceptions here or leave it to the func?

Reply:
My thinking was that we'd want to abort the migration on exception, but you're right, I should make that explicit so the contract is well-defined.